packages/smithy/smithy_aws/lib/src/http/interceptors/with_checksum.dart

```diff
@@ -0,0 +1,108 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:aws_common/aws_common.dart';
+import 'package:convert/convert.dart';
+import 'package:crclib/catalog.dart';
+import 'package:crclib/crclib.dart';
+import 'package:crypto/crypto.dart';
+import 'package:smithy/smithy.dart';
+
+class _CrcValueToHeaderConverter extends Converter<CrcValue, String> {
+  const _CrcValueToHeaderConverter();
+
+  @override
+  String convert(CrcValue input) {
+    return base64Encode(hex.decode(input.toRadixString(16)));
+  }
+
+  @override
+  Sink<CrcValue> startChunkedConversion(Sink<String> sink) {
+    return ChunkedConversionSink.withCallback(
+      (value) => sink
+        ..add(convert(value.last))
+        ..close(),
+    );
+  }
+}
+
+class _DigestToHeaderConverter extends Converter<Digest, String> {
+  const _DigestToHeaderConverter();
+
+  @override
+  String convert(Digest input) {
+    return base64Encode(input.bytes);
+  }
+
+  @override
+  Sink<Digest> startChunkedConversion(Sink<String> sink) {
+    return ChunkedConversionSink.withCallback(
+      (value) => sink
+        ..add(convert(value.last))
+        ..close(),
+    );
+  }
+}
+
+/// {@template smithy_aws.interceptors.with_checksum}
+/// Intercepts HTTP requests to include a content hash in the headers.
+///
+/// See: https://awslabs.github.io/smithy/2.0/aws/aws-core.html#aws-protocols-httpchecksum-trait
+/// {@endtemplate}
+class WithChecksum extends HttpRequestInterceptor {
+  /// {@macro smithy_aws.interceptors.with_checksum}
+  const WithChecksum([String? algorithm]) : _algorithm = algorithm ?? 'MD5';
+
+  final String _algorithm;
```

Comment:
Can the type of this field be `ChecksumAlgorithm` enum ?

Reply:
Unfortunately, no, with the way Smithy works, since the checksumAlgorithm needs to be part of the input and the input has to be fully defined in Smithy and is unique to that service.

The only way around it would be for Smithy to define a `ChecksumAlgorithm` type as a core type which could be used here, but it doesn't do that.